### PR TITLE
[Cognitive Services] Add completer for  `--sku`, `--kind`, `--location`

### PR DESCRIPTION
**Pull request by yangyuan**
Solves https://github.com/Azure/azure-cli/issues/7157

--

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py [hunk 1]
```diff
@@ -69,13 +74,64 @@ def _is_suspected_json(string):
...
+@Completer
+def location_completer(cmd, prefix, namespace, **kwargs):  # pylint: disable=unused-argument
+    names = set(sum([x.locations for x in _sku_filter(cmd, namespace)], []))
+    return [x.lower() for x in sorted(list(names))]
+
+
+def get_cognitiveservices_location_type(ctx):
+    _location_type = get_location_type(ctx)
+    _location_type.update(completer=location_completer)
+    return _location_type
```
Comment (williexu) [inline src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: No need to define a new arg_type; just use `get_location_type`.

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py [hunk 2]
```diff
@@ -69,13 +74,64 @@ def _is_suspected_json(string):
...
+@Completer
+def kind_completer(cmd, prefix, namespace, **kwargs):  # pylint: disable=unused-argument
+    kinds = set([x.kind for x in _sku_filter(cmd, namespace)])
+    return sorted(list(kinds))
+
+
+@Completer
+def location_completer(cmd, prefix, namespace, **kwargs):  # pylint: disable=unused-argument
+    names = set(sum([x.locations for x in _sku_filter(cmd, namespace)], []))
+    return [x.lower() for x in sorted(list(names))]
```
Comment (williexu) [inline src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: get_location_type already contains a completer: https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/commands/parameters.py#L247
Filtering is done in the CLI through argparse.

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py [hunk 3]
```diff
@@ -69,13 +74,64 @@ def _is_suspected_json(string):
...
-        c.argument('resource_group_name', arg_type=resource_group_name_type)
-        c.argument('sku_name', options_list=['--sku'], help='the Sku of cognitive services account')
-        c.argument('kind', help='the API name of cognitive services account')
+        c.argument('location', arg_type=get_cognitiveservices_location_type(self.cli_ctx))
+        c.argument('resource_group_name', arg_type=resource_group_name_type,
+                   completer=get_resource_group_completion_list)
+        c.argument('sku_name', options_list=['--sku'], help='the Sku of cognitive services account',
+                   completer=sku_name_completer)
+        c.argument('kind', help='the API name of cognitive services account',
+                   completer=kind_completer)
```
Comment (williexu) [inline src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: Best way to register enums is through `get_enum_type`: https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#registering-enums

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py [hunk 4]
```diff
@@ -69,13 +74,64 @@ def _is_suspected_json(string):
...
 def load_arguments(self, _):
     with self.argument_context('cognitiveservices') as c:
         c.argument('account_name', arg_type=name_arg_type, help='cognitive service account name',
                    completer=get_resource_name_completion_list('Microsoft.CognitiveServices/accounts'))
-        c.argument('resource_group_name', arg_type=resource_group_name_type)
-        c.argument('sku_name', options_list=['--sku'], help='the Sku of cognitive services account')
-        c.argument('kind', help='the API name of cognitive services account')
+        c.argument('location', arg_type=get_cognitiveservices_location_type(self.cli_ctx))
+        c.argument('resource_group_name', arg_type=resource_group_name_type,
+                   completer=get_resource_group_completion_list)
```
Comment (williexu) [inline src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: `resource_group_name_type` already contains the completer: https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/commands/parameters.py#L236

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 1 above)
Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: In here I need to override the default completer from the get_location_type.
So I guess I have to define a new type.

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 2 above)
Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: In here I want to give better experiences: only show locations available for the --kind and --sku (if specified)

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 4 above)
Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: okay, will remove this.

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 3 above)
Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: For cognitive services --sku and --kind can not simply be enums.

I explained to @tjprescott before.
For different subscriptions, the available skus and kinds (and their valid combinations) are different.
Cognitive Services contains a set of services, some of them have unique SKU models.

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 2 above)
Comment (williexu) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: I see

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py (hunk 1 above)
Comment (williexu) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: If you use "get_location_type" and just use the `completer` kwarg, the completer you specify will override the one in the type.

Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]:
```
def get_location_type(cli_ctx):
    location_type = CLIArgumentType(
        options_list=('--location', '-l'),
        completer=get_location_completion_list,
        type=get_location_name_type(cli_ctx),
        help="Location. You can configure the default location using `az configure --defaults location=<location>`",
        metavar='LOCATION',
        configured_default='location')
    return location_type
```

This is the definition of get_location_type(), it doesn't taking `kwarg`.
Did I miss something?

Comment (williexu) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: I mean when registering the parameter:
`c.argument('location', get_location_type(self.cli_ctx), completer=<>)`

Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: cool, change made as suggested.

File: src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py [hunk 5]
```diff
@@ -69,13 +73,58 @@ def _is_suspected_json(string):
...
+
+@Completer
+def kind_completer(cmd, prefix, namespace, **kwargs):  # pylint: disable=unused-argument
+    kinds = set([x.kind for x in _sku_filter(cmd, namespace)])
+    return sorted(list(kinds))
+
+
+@Completer
+def location_completer(cmd, prefix, namespace, **kwargs):  # pylint: disable=unused-argument
+    names = set(sum([x.locations for x in _sku_filter(cmd, namespace)], []))
```
Comment (williexu) [inline src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: `set([location for x in _sku_filter(cmd, namespace) for location in x.locations])` to be efficient

Comment (yangyuan) [reply src/command_modules/azure-cli-cognitiveservices/azure/cli/command_modules/cognitiveservices/_params.py]: ahh, true!

**Pull request merged**